gdscripts/console.gd

```diff
@@ -0,0 +1,391 @@
+            if command.contains(line_edit.text):
+                var index : int = command.find(line_edit.text)
+                if index <= prev_index:
+                    suggestions.push_front(command)
+                else:
+                    suggestions.push_back(command)
+                prev_index = index
+        autocomplete()
+
+
+func reset_autocomplete() -> void:
+    suggestions.clear()
+    current_suggest = 0
+    suggesting = false
+
+
+func toggle_size() -> void:
+    if (control.anchor_bottom == 1.0):
+        control.anchor_bottom = 1.9
+    else:
+        control.anchor_bottom = 1.0
+
+
+func disable():
+    enabled = false
+    toggle_console() # Ensure hidden if opened
+
+
+func enable():
+    enabled = true
+
+
+func toggle_console() -> void:
+    set_visible(!control.visible)
+
+func set_visible(visible:bool) -> void:
+    if not control or not line_edit:
+        return
+    if (enabled):
+        control.visible = visible
+    else:
+        control.visible = false
+
+    if (control.visible):
+        was_paused_already = get_tree().paused
+        get_tree().paused = was_paused_already || pause_enabled
+        line_edit.grab_focus()
+        console_opened.emit()
+    else:
+        control.anchor_bottom = 1.0
+        scroll_to_bottom()
+        reset_autocomplete()
+        if (pause_enabled && !was_paused_already):
+            get_tree().paused = false
+        console_closed.emit()
+
+
+func is_visible():
+    return control.visible
+
+
+func scroll_to_bottom() -> void:
+    var scroll: ScrollBar = rich_label.get_v_scroll_bar()
+    scroll.value = scroll.max_value - scroll.page
+
+
+func print_line(text : String, print_godot := false) -> void:
+    if (!rich_label): # Tried to print something before the console was loaded.
+        call_deferred("print_line", text)
+    else:
+        rich_label.append_text(text)
+        rich_label.append_text("\n")
+        if (print_godot):
+            print(text)
+
+
+func parse_line_input(text : String) -> PackedStringArray:
+    var out_array : PackedStringArray
+    var first_char := true
+    var in_quotes := false
+    var escaped := false
+    var token : String
+    for c in text:
+        if (c == '\\'):
+            escaped = true
+            continue
+        elif (escaped):
+            if (c == 'n'):
+                c = '\n'
+            elif (c == 't'):
+                c = '\t'
+            elif (c == 'r'):
+                c = '\r'
+            elif (c == 'a'):
+                c = '\a'
+            elif (c == 'b'):
+                c = '\b'
+            elif (c == 'f'):
+                c = '\f'
+            escaped = false
+        elif (c == '\"'):
+            in_quotes = !in_quotes
+            continue
+        elif (c == ' ' || c == '\t'):
+            if (!in_quotes):
+                if token:
+                    out_array.push_back(token)
+                token = ""
+                continue
+        token += c
```

Comment:
Don't we have a simpler way to achieve this?
Maybe mapping in a dictionary instead of a set of if/else?

```suggestion
    var mappings = {"n": "\n", etc...}
    for c in text:
        if (c == '\\'):
            escaped = true
            continue
        elif (escaped):
            var map = mappings[c]
            if map !== null:
                c = map
            escaped = false
        elif (c == '\"'):
            in_quotes = !in_quotes
            continue
        elif (c == ' ' || c == '\t'):
            if (!in_quotes):
                if token:
                    out_array.push_back(token)
                token = ""
                continue
        token += c
```

Reply:
As stated in line 3: `Based on https://github.com/jitspoe/godot-console (MIT)`

This is copy-paste from this addon with minor fixes. The godot-console is badly implemented. But I think  we don't have time to write console script from the scratch for now. I suggest moving this to a new issue about reimplementation of `console.gd`.